src/esp32/i2c.rs

```diff
@@ -0,0 +1,187 @@
+        }
+    }
+}
+
+impl TryFrom<&Kind> for Esp32I2cConfig {
+    type Error = AttributeError;
+    fn try_from(value: &Kind) -> Result<Self, Self::Error> {
+        match value {
+            Kind::StructValueStatic(v) => {
+                if !v.contains_key("name") {
+                    return Err(AttributeError::KeyNotFound);
+                }
+                let name = v.get("name").unwrap().try_into()?;
+                if !v.contains_key("bus") {
+                    return Err(AttributeError::KeyNotFound);
+                }
+                let bus = v.get("bus").unwrap().try_into()?;
+                let mut data_pin = 11;
+                if v.contains_key("data_pin") {
+                    data_pin = v.get("data_pin").unwrap().try_into()?;
+                }
+                let mut clock_pin = 6;
+                if v.contains_key("clock_pin") {
+                    clock_pin = v.get("clock_pin").unwrap().try_into()?;
+                }
+                let mut baudrate_hz: u32 = 1000000;
+                if v.contains_key("baudrate_hz") {
+                    baudrate_hz = v.get("baudrate").unwrap().try_into()?;
+                }
+                let mut timeout_ns: u32 = 0;
+                if v.contains_key("timeout_ns") {
+                    timeout_ns = v.get("timeout_ns").unwrap().try_into()?;
+                }
+                Ok(Self {
+                    name,
+                    bus,
+                    baudrate_hz,
+                    timeout_ns,
+                    data_pin,
+                    clock_pin,
+                })
+            }
+            _ => Err(AttributeError::ConversionImpossibleError),
+        }
+    }
+}
+
+pub struct Esp32I2C<'a> {
+    name: String,
+    driver: I2cDriver<'a>,
+    timeout_ns: u32,
+}
+
+impl<'a> Esp32I2C<'a> {
+    pub fn new_from_config(conf: Esp32I2cConfig) -> anyhow::Result<Self> {
+        let name = conf.name.to_string();
+        let timeout_ns = conf.timeout_ns;
+        let peripherals = Peripherals::take().unwrap();
+        let sda = unsafe { AnyIOPin::new(conf.data_pin) };
+        let scl = unsafe { AnyIOPin::new(conf.clock_pin) };
+        let driver_conf = I2cConfig::from(conf);
+
+        match conf.bus {
+            "i2c0" => {
+                let driver = I2cDriver::new(peripherals.i2c0, sda, scl, &driver_conf)?;
+                Ok(Esp32I2C {
+                    name,
+                    driver,
+                    timeout_ns,
+                })
+            }
+            "i2c1" => {
+                let driver = I2cDriver::new(peripherals.i2c1, sda, scl, &driver_conf)?;
+                Ok(Esp32I2C {
+                    name,
+                    driver,
+                    timeout_ns,
+                })
+            }
+            _ => anyhow::bail!("only i2c0 or i2c1 supported, i2c bus must match either value"),
+        }
+    }
+}
+
+impl<'a> I2CHandle<u8> for Esp32I2C<'a> {
```

Comment:
Are you planning to remove the template on I2CHandle for the address type?

Reply:
oh, I forgot I accommodated that. I'll remove it, but out of curiosity do you know of any processors that use 10-bit addresses for I2C?

Reply:
not to mu knowledge